libs/community/langchain_community/document_loaders/batch_loader.py

```diff
@@ -0,0 +1,288 @@
+import multiprocessing
+from concurrent.futures import ThreadPoolExecutor
+from functools import partial
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Iterable,
+    List,
+    Literal,
+    TypeVar,
+)
+
+from langchain_core.documents import Document
+
+from langchain_community.document_loaders.base import BaseLoader
+
+T = TypeVar("T")
+
+
+def _make_iterator_list(
+    iterable: Iterable[T],
+    args_length: int,
+    show_progress: bool = False,
+) -> List[T]:
+    """Make an iterator from an iterable. If show_progress is True, will
+    return a tqdm iterator.
+
+    Args:
+        iterable (Iterable[T]): any iterable
+        args_length (int): length of the iterable
+        show_progress (bool, optional): If true, will return a tqdm iterator.
+            Defaults to False.
+
+    Raises:
+        ImportError: if tqdm is not installed and show_progress is True
+    """
+    if show_progress:
+        try:
+            from tqdm import tqdm
+        except ImportError as err:
+            raise ImportError(
+                "You must install tqdm to use show_progress=True."
+                "You can install tqdm with `pip install tqdm`."
+            ) from err
+        return list(tqdm(iterable, total=args_length))
+    return list(iterable)
+
+
+def _load_loader(
+    loader_args: Dict[str, Any], loader_callable: Callable[..., BaseLoader]
+) -> List[Document]:
+    """Function used to instantiate the loader and load the data.
+
+    Args:
+        loader_args (Dict[str, Any]): list of loader_args used
+            to initialize the loader. The keys of the dict are the name of
+            the arguments and the values are the list of values for each.
+        loader_callable (Callable[..., BaseLoader]): callable
+            used to initialize the loader (class or init function )
+
+    Returns:
+        List[Document]: list of loaded documents
+    """
+    loader = loader_callable(**loader_args)
+    return loader.load()
+
+
+Methods = Literal["sequential", "thread", "process", "async"]
+
+
+class BatchLoader(BaseLoader):
+    """
+    Meta loader use to load multiple data from a loader either in parallel
+    or sequentially.
+
+    Example:
+
+    Using a class that is instantiate with __init__
+    ```
+    file_path_list : List[str] = [...]
+    loader_files = BatchLoader(
+        TextLoader,
+        {
+            "file_path": file_path_list
+        },
+        method="process",
+        max_workers=4,
+    )
+    ```
+    Using a static method to initialize the loader:
+    ```
+    youtube_video_urls : List[str] = [...]
+    loader_youtube = BatchLoader(
+        YoutubeLoader.from_url,
+        {
+            "url": youtube_video_urls
+        },
+    )
+    ```
+
+    Then you can load your data:
+    ```
+    documents = loader_files.load()
+    ```
+    """
+
+    def __init__(
+        self,
+        loader_initializer_callable: Callable[..., BaseLoader],
+        init_loader_args: Dict[str, List[Any]],
```

Comment:
Wouldn't it be easier to just pass a list of loaders ?

Reply:
You mean a list of already instantiated loaders ? Indeed this could simplify the code of the batch loader and allow users to load data from multiple loaders using 1 batch loader. 
But I'm not sure if, from a developer perspective, it would be a better experience or not.
User's code would be:
```python
# Propisition 1
BatchLoader(
    YoutubeLoader.from_youtube_url,
    {"youtube_url": youtube_videos}
)

# Proposition 2
loaders = [YoutubeLoader.from_youtube_url(url) for url in youtube_videos]
BatchLoader(loaders)
```

I do not have strong conviction on what could be the best way of doing that, if you prefer I can simplify this of course